bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/ScriptTransformationServiceFactory.java

```diff
@@ -0,0 +1,104 @@
+
+import javax.script.ScriptEngine;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.automation.module.script.internal.ScriptEngineFactoryHelper;
+import org.openhab.core.transform.TransformationService;
+import org.osgi.service.component.ComponentFactory;
+import org.osgi.service.component.ComponentInstance;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * The {@link ScriptTransformationServiceFactory} registers a {@link ScriptTransformationService}
+ * for each newly added script engine.
+ *
+ * @author Jimmy Tanagra - Initial contribution
+ */
+@Component(immediate = true, service = { ScriptTransformationServiceFactory.class })
+@NonNullByDefault
+public class ScriptTransformationServiceFactory {
+
+    private final ComponentFactory<ScriptTransformationService> scriptTransformationFactory;
+
+    private final Map<String, ComponentInstance<ScriptTransformationService>> scriptTransformations = new ConcurrentHashMap<>();
+
+    @Activate
+    public ScriptTransformationServiceFactory(
+            @Reference(target = "(component.factory=org.openhab.core.automation.module.script.transformation.factory)") ComponentFactory<ScriptTransformationService> factory) {
+        this.scriptTransformationFactory = factory;
+    }
+
+    @Deactivate
+    public void deactivate() {
+        scriptTransformations.values().forEach(this::unregisterService);
+        scriptTransformations.clear();
+    }
+
+    /**
+     * As {@link ScriptEngineFactory}s are added/removed, this method will cache all available script types
+     * and registers a transformation service for the script engine.
+     */
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void setScriptEngineFactory(ScriptEngineFactory engineFactory) {
+        Optional<String> scriptType = ScriptEngineFactoryHelper.getPreferredExtension(engineFactory);
+        if (scriptType.isEmpty()) {
+            return;
+        }
+
+        scriptTransformations.computeIfAbsent(scriptType.get(), type -> {
+            ScriptEngine scriptEngine = engineFactory.createScriptEngine(type);
+            if (scriptEngine == null) {
+                return null;
+            }
+            String languageName = ScriptEngineFactoryHelper.getLanguageName(scriptEngine.getFactory());
+            Dictionary<String, Object> properties = new Hashtable<>();
+            String name = type.toUpperCase();
+            properties.put(TransformationService.SERVICE_PROPERTY_NAME, name);
+            properties.put(TransformationService.SERVICE_PROPERTY_LABEL, name + " - " + languageName);
```

Comment:
I would prefer `SCRIPT Rule DSL (v1)` or `SCRIPT ECMAScript 262 Edition`. This also groups them nicely with your other PR.

Reply:
Yes this looks nicer! changed as suggested

Reply:
I’d like to have the JS and NASHORNJS names still there, because they are mich easier to differ then the ECMAScript Versions, but I am fine with putting a SCRIPT in front of them.

Reply:
That makes the label very long and duplicates the information for all other languages. Can we modify what is returned by the script engine so it returns „ECMAScript (Nashorn, v5)“ and „ECMAScript (Graal, v11)“ or similar?

Reply:
I don’t think this is possible, because at least I’ve never noticed GraalJS providing the information that it is Graal. 

However looking at the UI for creating new scripts, I think it would be fine if the config description contains the MIME type and UI can display the MIME type. And don’t change the label here.

Reply:
How about `SCRIPT + " "  scriptType + " " + Language Name (without the extra version number)`

- SCRIPT DSL Rule DSL
- SCRIPT GROOVY Groovy
- SCRIPT JS ECMAScript
- SCRIPT NASHORNJS ECMAScript
- SCRIPT PY Python
- SCRIPT RB Ruby

Reply:
> I think it would be fine if the config description contains the MIME type and UI can display the MIME type. And don’t change the label here.

Could you clarify this for me please?

Reply:
That would be fine, but how can you do that without putting these labels in core code?

Reply:
```java
    public static String getLanguageName(javax.script.ScriptEngineFactory factory) {
        String languageName = factory.getLanguageName();
        String languageVersion = factory.getLanguageVersion();
        if (languageName.equals("ECMAScript")) {
            if (factory.getLanguageVersion().contains(" 5.1")) {
                languageName = "NashornJS";
            } else {
                languageName = "JavaScript";
            }
        }
        return String.format("%s (%s)", languageName.substring(0, 1).toUpperCase() + languageName.substring(1),
                languageVersion);
    }
 ```

Reply:
No, don‘t implement something in core that is specific to single add-ons.

Reply:
Alternatively we change the corresponding graalvs and nashorn bundles to somehow return those values, but that involves creating a proxy javax.script.ScriptEngineFactory unless there's a better way to do that.

Reply:
FYI, this is what `GraalJSEngineFactory` which is implementing [`javax.script.ScriptEngineFactory`](https://docs.oracle.com/en/java/javase/17/docs/api/java.scripting/javax/script/ScriptEngineFactory.html):

https://github.com/oracle/graaljs/blob/a314832c8fc1e7f06244d9afb0553900f4d5a638/graal-js/src/com.oracle.truffle.js.scriptengine/src/com/oracle/truffle/js/scriptengine/GraalJSEngineFactory.java#L55-L63

I was wrong when I wrote that I never noticed that Graal provides some information that it is Graal, sorry.

Reply:
> Alternatively we change the corresponding graaljs and nashorn bundles to somehow return those values, but that involves creating a proxy javax.script.ScriptEngineFactory unless there's a better way to do that.

Both GraalJS and NashornJS addons already have use their own ScriptEngineFactory with method overrides to provide e.g. additional MIME types. See https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.automation.jsscriptingnashorn/src/main/java/org/openhab/automation/jsscriptingnashorn/internal/NashornScriptEngineFactory.java and https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/GraalJSScriptEngineFactory.java.

Therefore let‘s just override the `getLanguageName` method there and use the languageName as in the current state of this PR. Therefore this PR should be finished.

Reply:
Using getEngineName

<img width="632" alt="image" src="https://user-images.githubusercontent.com/2554958/228363660-f9df716b-e5b1-497b-a9a6-c0bdb1cc5391.png">

<img width="443" alt="image" src="https://user-images.githubusercontent.com/2554958/228363734-0ade146d-378f-4013-914c-f49e94c57dbb.png">

Reply:
Using `getEngineName` IMO is too long. I‘d go for the languageName and overwrite the `getLanguageName` methods in the JS bundles.

Reply:
> Using `getEngineName` IMO is too long. I‘d go for the languageName and overwrite the `getLanguageName` methods in the JS bundles.

I agree. If this is our decision, then there's no further changes needed for this PR?

It will look like this now, before changes in the JS bundles
<img width="408" alt="image" src="https://user-images.githubusercontent.com/2554958/228369715-6799184c-054b-4539-850b-5fcc4a5f95e6.png">

Reply:
Looks good 👍 

From my POV, this PR is finished.